File: java/sort/qsort_basic/QSortBasic.java
import java.lang.*;

public class QSortBasic {
  private static void swap (int[] arr, final int i, final int j) {
    final int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  private static void qsort (int[] arr, final int lo, final int hi) {
    if (lo >= hi) return;

    final int pIndex = ((int) (Math.random() * (hi - lo))) + lo,
              pVal   = arr[pIndex];
    int       swapPt = lo;

    swap(arr, lo, pIndex);
    for (int i = lo + 1; i <= hi; i++)
      if (arr[i] < pVal) swap(arr, i, ++swapPt);
    swap(arr, lo, swapPt);
    
    qsort(arr, lo, swapPt - 1);
    qsort(arr, swapPt + 1, hi);
  }

  public static void main(String[] args) {
    int[] testArr = {5,2,78,5,0,13,6};
    for (int i : testArr) System.out.printf("%d%c", i, ' ');
    System.out.println();

    qsort(testArr, 0, testArr.length - 1);
    for (int i : testArr) System.out.printf("%d%c", i, ' ');
  }
}

File: javascript/sort/qsort_basic.js
'use strict';

Array.prototype.swap = function (i, j) {
  const temp = this[i];
  this[i] = this[j];
  this[j] = temp;
}

function qsort (arr, lo=0, hi=arr.length - 1) {
  if (lo >= hi) return;
  
  arr.swap(lo, Math.floor(Math.random() * (hi - lo)) + lo);

  const pivot     = arr[lo];
  let   swapPoint = lo;

  for (let i = lo + 1; i <= hi; i++) {
    if (arr[i] < pivot) arr.swap(i, ++swapPoint);
  }
  arr.swap(lo, swapPoint);

  qsort(arr, lo, swapPoint - 1);
  qsort(arr, swapPoint + 1, hi);

  // return arr;
}

const testArr = [5,2,78,5,0,13,6];
qsort(testArr);
console.log(testArr);